Preferences Dialog Pattern on site (#2965)
jpmorganchase/salt-ds: site/docs/patterns/preferences-dialog.mdx

---
title: Preferences Dialog
layout: DetailPattern
aliases:
  - /salt/patterns/preferences-dialog
data:
  resources:
    [

>       { href: "https://www.saltdesignsystem.com/salt/components/navigation-item/examples", label: "Navigation item", internal: false },

Reviewer: ```suggestion
```

>       { href: "https://www.figma.com/file/N6cSYycc4z6XkCjVvc1kIA/Preferences-Dialog-Pattern?type=design&node-id=157-40560&mode=design&t=eq3OapLjaquXm3UP-0", label: "Preferences Dialog stickhersheet", internal: true },

Reviewer: ```suggestion
      { href: "https://www.figma.com/file/N6cSYycc4z6XkCjVvc1kIA/Preferences-Dialog-Pattern?type=design&node-id=157-40560&mode=design&t=eq3OapLjaquXm3UP-0", label: "Preferences Dialog stickersheet", internal: true },
```

    ]
---

A preferences dialog allows you to display, group and navigate application preferences within a responsive modal dialog window. This pattern introduces guidelines to make the preferences dialog consistent and usable.
  
  <Diagram

>     src="/img/patterns/preferences-dialog/preferences-dialog.png"

Reviewer 2: the bottom box shadow seems to be cut off in this image

Author: @navkaur76 do you think you can add a bit more space under the dialog so the shadow would not look cut? This goes for most of the images.

Author: sorted. Thanks @navkaur76

    alt="Preferences Dialog example"
    border={false}
  />

## When to use

> You should ideally house any editable settings within a centralized preferences dialog to display and manage user preferences across a platform, application or module. Displaying preferences should be contextual, i.e., the dialog should contain all preferences but depending on the entry point to the user interface, the preferences shown to you in the first instance will vary. 

Reviewer 3: ```suggestion
Any preference settings in your platform or application should ideally found within a single centralised preferences dialog. These settings could be global, application-wide settings or settings local to a specific window or widget (such as a grid or comments feed). If accessing from a specific window, the relevant settings should immediately be displayed when the preferences dialog opens. 
```

> Any preference settings in your platform or application should ideally be found within a single centralised preferences dialog. These settings could be global, application-wide settings or settings local to a specific window or widget (such as a grid or comments feed). If accessing from a specific window, the relevant settings should immediately be displayed when the preferences dialog opens.

Reviewer: ```suggestion
Any preference settings in your platform or application should ideally be found within a single centralized preferences dialog. These settings could be global, application-wide settings or settings local to a specific window or widget (such as a grid or comments feed). If accessing from a specific window, the relevant settings should immediately be displayed when the preferences dialog opens.
```

Furthermore, we recommend implementing consistent behavior around applying preferences, whether applying changes immediately or requiring the user to confirm any changes using the button bar actions. 

## How to build

### Anatomy

A preferences dialog comprises four key features: 

> 1. **Dialog:** a window with a content area that opens over the application content, focusing the user’s attention on a particular task or piece of information

Reviewer: are we happy with calling it a window?

Reviewer 3: We refer to dialog as a "window" in the component docs too... I would think "modal" is more fitting? Maybe something to check with Mike

> 2. **Navigation item:** a list of links that allows users to navigate between distinct panels of content

Reviewer: Navigation item is a single link not a list.

> 3. **Button bar:** displays possible actions for users to take when they reach the end of a task. Use the button bar pattern to understand how to implement actionable interactions within dialogs across your applications

Reviewer 3: Yep, there's links missing. Each of the key components here should be linked out to their relevant component pages on the site, e.g. Dialog > dialog component page. I also have done exactly this in figma if you need to check where else we are linking out to stuff, or just give me a shout @alycrys

> 4. [Parent-child layout](../components/parent-child-layout): positions the navigation and content within a preference dialog

Reviewer 4: i think we dont have this component/link live yet - is it better not to link it until we have it? conscious it leads you to an error that says Page broken now

Author: I have a PR ready for review https://github.com/jpmorganchase/salt-ds/pull/2950. Once it gets merged, this will fix itself

  <Diagram
    src="/img/patterns/preferences-dialog/anatomy.png"
    alt="Preferences Dialog navigation and content."
    border={false}
  />

### Layout construction

<ImageSwitcher
  images={[
    {
      src: "/img/patterns/preferences-dialog/layout.png",

>       alt: "Example showing Preferences Dialog's layout construction.",

Reviewer: Duplicate alt text could be confusing to screen readers.

    },
    {
      src: "/img/patterns/preferences-dialog/layout-grid.png",
      alt: "Example showing Preferences Dialog's layout construction with grid.",
    },
  ]}
  label="Layout grid"
/>

<ExampleContainer type="positive" customPillText="Best practices">

>   In the [Parent-child layout](../components/parent-child-layout), the parent container (containing the navigation) typically remains fixed in width when the user resizes the browser. You can apply an independent layout grid to align the content with the child container. This will enable it to adapt responsively based on the viewport size.

Reviewer 3: ```suggestion
  In the [parent-child layout](../components/parent-child-layout), the parent container (containing the navigation) typically remains fixed in width when the user resizes the browser. You can apply an independent layout grid to align the content with the child container. This will enable it to adapt responsively based on the viewport size.
```

  To implement a responsive experience for preferences dialogs, refer to the layout guidance in the [Navigation pattern](./navigation).

</ExampleContainer>

## Compositions

### Navigation

You can use navigation items to switch between pages of grouped preferences. 

  <Diagram
    src="/img/patterns/preferences-dialog/navigation.png"

>     alt="Example of navigation."

Reviewer: Can we add more context here?

    border
  />

<ExampleContainer type="positive" customPillText="Best practices">
  Limit nesting to no more than two levels, if possible. A deep hierarchy can make it difficult for users to discover or locate specific preferences.
</ExampleContainer>

### Scrolling

> We recommend that scrolling affordance should only take effect if content in the Parent Child layout exceeds the visible content area.

Reviewer 3: ```suggestion
We recommend that scrolling affordance should only take effect if content in the parent-child layout exceeds the visible content area.
```

The visual indicators are the two separator lines above and below the parent-child layout. Apply the shadow-scroll token for the container shadow. 

  <Diagram
    src="/img/patterns/preferences-dialog/scrolling.png"

>     alt="Example of scrolling."

Reviewer: and here

<Diagram
  src="/img/patterns/preferences-dialog/applying-preferences.png"
  alt="Applying preferences."
  border
/>

<ExampleContainer type="positive" customPillText="Best practices">
  Display core actions to the right of a dialog. Refer to the [button bar pattern](./button-bar) for further guidance on how to display user actions within dialogs. 
</ExampleContainer>

### Live updates

If your application can listen for changes and take immediate effect, you don't need to display a button bar across a preferences dialog. A close button in the top-right corner of the dialog should be on display instead to allow the user to close the dialog after editing any preferences.

<Diagram
  src="/img/patterns/preferences-dialog/live-updates.png"
  alt="Live updates."
  border
/>

## Secondary actions

We recommend configuring the button bar to display secondary actions to the left such as **Import** and **Export** if you need to enable additional actions.

<Diagram
  src="/img/patterns/preferences-dialog/secondary-actions.png"

>   alt="Live updates."

Reviewer: This alt text is wrong

  border
/>

## Responsiveness

> The preferences dialog is responsive to the width of the application viewport and respond dynamically to the Salt breakpoints. 

Reviewer 3: ```suggestion
The preferences dialog is responsive to the width of the application viewport and responds dynamically to the Salt breakpoints. 
```

Once a preference dialog hits a certain breakpoint, you can configure the layout to either: 

1. Collapse to allow more room for content.
2. Change the orientation to stack the button bar.

### Collapsed navigation

> When the navigation collapses, the preferences content displays across the entire width of the dialog. To select and view alternative preference types, users should use the back chevron button to navigate back to the vertical navigation. The vertical navigation view also spans the dialog width. A collapsed layout ensures improved usability of dialog content in smaller screen sizes or application resolutions.

Reviewer: We switch between "displays across" and "spans" in the same sentence. Should we just use "spans"

Reviewer 3: ```suggestion
When the navigation collapses, the preferences content spans the entire width of the dialog. To select and view alternative preference types, users should use the back chevron button to navigate back to the vertical navigation. The vertical navigation view also spans the dialog width. A collapsed layout ensures improved usability of dialog content in smaller screen sizes or application resolutions.
```
